src/slot.rs

```diff
@@ -642,102 +648,238 @@ impl Stats {
-    ) -> Self {
-        let (slot_tx, slot_rx) = mpsc::channel(config.max_count);
-        let mut set = Self {
-            pool_name,
-            name,
-            backend,
-            config,
-            wanted_count,
-            backend_connector,
-            stats,
-            failure_window,
-            rx,
-            terminate_rx,
-            status_tx,
-            slot_tx,
-            slot_rx,
-            slots: BTreeMap::new(),
-            next_slot_id: SlotId::first(set_id),
+impl<Conn: Connection> Slots<Conn> {
+    #[instrument(
+        level = "trace",
+        skip(self),
+        err,
+        name = "Slots::claim",
+        fields(name = ?self.name),
+    )]
+    fn claim(&mut self, id: ClaimId) -> Result<claim::Handle<Conn>, Error> {
+        #[cfg(feature = "probes")]
+        probes::slot__set__claim__start!(|| (
+            self.pool_name.as_str(),
+            id.0,
+            self.backend.address.to_string()
+        ));
+
+        // Before we vend out the slot's connection to a client, make sure that
+        // we have space to take it back once they're done with it.
+        let Ok(permit) = self.slot_tx.clone().try_reserve_owned() else {
+            event!(Level::TRACE, "Could not reserve slot_tx permit");
+
+            #[cfg(feature = "probes")]
+            probes::slot__set__claim__failed!(|| (
+                self.pool_name.as_str(),
+                id.0,
+                "Could not reserve slot_tx permit; all slots used"
+            ));
+
+            // This is more of an "all slots in-use" error,
+            // but it should look the same to clients.
+            return Err(Error::NoSlotsReady);
         };
-        set.set_wanted_count(wanted_count);
-        set
+
+        let Some(handle) = self.take_connected_unclaimed_slot(permit, id) else {
+            event!(Level::TRACE, "Failed to take unclaimed slot");
+
+            #[cfg(feature = "probes")]
+            probes::slot__set__claim__failed!(|| (
+                self.pool_name.as_str(),
+                id.0,
+                "No unclaimed slots"
+            ));
+            return Err(Error::NoSlotsReady);
+        };
+
+        #[cfg(feature = "probes")]
+        probes::slot__set__claim__done!(|| (
+            self.pool_name.as_str(),
+            id.0,
+            handle.slot_id().as_u64()
+        ));
+
+        return Ok(handle);
     }
 
-    // Creates a new Slot, which always starts as "Connecting", and spawn a task
-    // to actually connect to the backend and monitor slot health.
+    // Borrows a connection out of the first unclaimed slot.
+    //
+    // Returns a Handle which has enough context to put the claim back,
+    // once it's dropped by the client.
     #[instrument(
-        skip(self)
+        skip(self, permit)
         fields(pool_name = %self.pool_name),
-        name = "SetWorker::create_slot"
+        name = "Slots::take_connected_unclaimed_slot"
     )]
+    fn take_connected_unclaimed_slot(
```

Comment:
This function has basically been moved from the `SetWorker` to this new struct, `Slots`